### PR TITLE
Coil_pertubation fix by adding Coils_from_gamma

### PR DESCRIPTION
**Pull request by eduardolneto**
Adding Coils_from_gamma in coils.py and modifying coil_pertubation.py to work with it. Also modifying simple_examples/create_pertubed_coils.py to work with the changes.

### CODE REVIEW

File: essos/coil_perturbation.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import jax.numpy as jnp
 from jax import jit, vmap
 from jaxtyping import Array, Float  # https://github.com/google/jaxtyping
-from essos.coils import Curves,apply_symmetries_to_gammas
+from essos.coils import Curves, Coils, Coils_from_gammas, apply_symmetries_to_gammas, fit_dofs_from_coils
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply essos/coil_perturbation.py]: Not a priority

File: essos/coil_perturbation.py [hunk 2]
```diff
@@ -205,70 +205,92 @@ def get_sample(self, deriv):
...
                 
     Returns:
-        The curves given as an input are modified and thus no return is done
+        A new perturbed curves object of the same type as the input.
+        The original input object is not modified.
+        
+    Note:
+        Statistical perturbations require disabling symmetry (nfp=1, stellsym=False)
+        since each coil gets an independent perturbation.
     """
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply essos/coil_perturbation.py:259]: Ok suggestion, but not a priority

File: essos/coil_perturbation.py [hunk 3]
```diff
@@ -205,70 +205,92 @@ def get_sample(self, deriv):
...
         perturbation = jax.vmap(sampler.draw_sample, in_axes=(0))(new_seeds)
-        gamma_perturbations = apply_symmetries_to_gammas(perturbation[:,0,:,:], curves.nfp, curves.stellsym)
-        gamma_perturbations_dash = apply_symmetries_to_gammas(perturbation[:,1,:,:], curves.nfp, curves.stellsym)
-        curves.gamma=curves.gamma + gamma_perturbations    
-        curves.gamma_dash=curves.gamma_dash + gamma_perturbations_dash                   
-    elif sampler.n_derivs == 2:
+        base_gamma = Curves(curves.dofs_curves, curves.n_segments, nfp=1, stellsym=False).gamma
+        perturbed_base_gamma = base_gamma + perturbation[:, 0, :, :]
+        dofs_new, _ = fit_dofs_from_coils(perturbed_base_gamma, curves.order, curves.n_segments)
+        curves.dofs_curves = dofs_new
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply essos/coil_perturbation.py]: This seems important

File: essos/coils.py [hunk 4]
```diff
@@ -814,4 +814,546 @@ def fit_dofs_from_coils(
...
+    
+    def to_json(self, filename: str):
+        """Save coils to JSON file"""
+        data = {
+            "n_segments": self.n_segments,
+            "nfp": self.nfp,
+            "stellsym": self.stellsym,
+            "dofs_gamma": self.dofs_gamma.tolist(),
+            "dofs_currents": self.dofs_currents.tolist(),
+        }
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply essos/coils.py]: This also seems important

File: essos/coils.py [hunk 5]
```diff
@@ -814,4 +814,546 @@ def fit_dofs_from_coils(
...
+    
+    # curvature property
+    @staticmethod
+    @jit
+    def compute_curvature(gammadash, gammadashdash):
+        return jnp.linalg.norm(jnp.cross(gammadash, gammadashdash, axis=1), axis=1) / jnp.linalg.norm(gammadash, axis=1)**3
+    
+    @property
+    def curvature(self):
+        return vmap(self.compute_curvature)(self.gamma_dash, self.gamma_dashdash)
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply essos/coils.py]: This is important to mimic the functionality that @EstevaoMGomes introduced

File: examples/simple_examples/create_perturbed_coils.py [hunk 6]
```diff
@@ -31,25 +30,27 @@
...
 key=0
 split_keys=jax.random.split(jax.random.key(key), num=2)
 #Add systematic error
-coils_sys = Coils(curves=curves, currents=[current_on_each_coil]*number_coils_per_half_field_period)
+coils_sys = CoilsFromGamma(base_gamma, currents=base_currents, nfp=number_of_field_periods, stellsym=True)
 perturb_curves_systematic(coils_sys, g, key=split_keys[0])
-# Add statistical error
-coils_stat = Coils(curves=curves, currents=[current_on_each_coil]*number_coils_per_half_field_period)
-perturb_curves_statistic(coils_stat, g, key=split_keys[1])
+# Add statistical error (returns a new object because now there are no symmtries in the perturbed coils so nfp and stellsym changes)
```
[the remark this answers is not in the document]
Comment (rogeriojorge) [reply examples/simple_examples/create_perturbed_coils.py:47]: Meh

**Review by rogeriojorge — COMMENTED** [review]
Mostly good with me, see the few specific comments I wrote above

File: essos/coils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eduardolneto) [reply essos/coils.py]: Changed!

File: essos/coils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eduardolneto) [reply essos/coils.py]: So I already solved this for using already optimized coils during alpha particle optimizations, but this change also included the saving of the scaling factors in order fot he coils to be recosntructed with the correct scaling of both fourier modes and currents. Therefore, I would add already the scaling changes I had inlcuded for the coils as well as the coils from a given set of axis function, and that would include all coils modifications?

File: essos/coil_perturbation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eduardolneto) [reply essos/coil_perturbation.py]: Changed.

**Comment by eduardolneto** [comment]
I solved the last comments and added scaling for Fourier Coil dofs as well as method CreateCoilsAroundAxis which will generate coils around a given axis or around the axis of a given surface.

File: essos/augmented_lagrangian.py [hunk 7]
```diff
@@ -15,150 +15,493 @@ class LagrangeMultiplier(NamedTuple):
...
+#     #     del params
+#     #     if model_lagrangian=='Standard' :
+#     #         return update_method(lagrange_params,updates,eta,omega,model_mu=model_mu,beta=beta,mu_max=mu_max,alpha=alpha,gamma=gamma,epsilon=epsilon,eta_tol=eta_tol,omega_tol=omega_tol), state
+#     #     elif model_lagrangian=='Squared' :
+#     #         return update_method_squared(lagrange_params,updates,eta,omega,model_mu=model_mu,beta=beta,mu_max=mu_max,alpha=alpha,gamma=gamma,epsilon=epsilon,eta_tol=eta_tol,omega_tol=omega_tol), state
+#     #     else:
+#     #         print('Lagrangian model not available please select Standard or Squared ')
+#     #         os._exit(0)              
+
+#     return optax.GradientTransformation(init_fn, update_fn)
```
Comment (EstevaoMGomes) [inline essos/augmented_lagrangian.py]: Having this much commented code is questionable. The commented code should be deleted, and an issue should be raised mentioning that this should be reimplemented from a given commit (with the hash)

File: essos/coils.py [hunk 8]
```diff
@@ -25,7 +25,15 @@ def __init__(self,
                  dofs:        jnp.ndarray,
                  n_segments:  int = 100,
                  nfp:         int = 1,
-                 stellsym:    bool = True):
+                 stellsym:    bool = True,
+                 scaling_type: int = 2,
+                 scaling_factor: float = 0,
+                 scale_fixed: float = 1.0):
```
Comment (EstevaoMGomes) [inline essos/coils.py]: Care has to be taken when adding more params (I guess static) to the Curves PyTree

File: essos/coils.py [hunk 9]
```diff
@@ -814,4 +1225,547 @@ def fit_dofs_from_coils(
         gamma_uni = _resample_closed_curve_uniform_batch(coils_gamma, n_segments)  # arclength (vmapped)
 
     dofs = _fit_real_fourier_batch(gamma_uni, order)  # rFFT-based fit
-    return dofs, gamma_uni
\ No newline at end of file
+    return dofs, gamma_uni
+
+class CoilsFromGamma:
+    """ Class to store coils from gamma (discretized curve coordinates) instead of Fourier coefficients
```
Comment (EstevaoMGomes) [inline essos/coils.py]: Maybe then call it `DiscretizedCoils`

**Review by EstevaoMGomes — COMMENTED** [review]
PR looks good; however, I did not test it, so it needs to be rechecked.
This is, nevertheless, a feat PR, so it should be merged only after [PR!35](https://github.com/uwplasma/ESSOS/pull/35), which is a fix PR.

**Comment by eduardolneto** [comment]
@EstevaoMGomes PR https://github.com/uwplasma/ESSOS/pull/35 Is trying to fix things in a wrong way for coils, surfaces, and augmented_lagrangian and arguably for objective_functions. These things were alreayd solved here and this should have been pushed before. Some of these for example avoid the necessity of some coils setter problems that are mentioned in https://github.com/uwplasma/ESSOS/pull/35 . So pushing the other one will conflict with this.

File: essos/coils.py (hunk 9 above)
Comment (eduardolneto) [reply essos/coils.py]: I mean the Coils cosntructed from Fourier coefficients would technically be SpectralCoils then to be consistent?

**Comment by rogeriojorge** [comment]
Hi @eduardolneto can you make this comment on PR #35 so that @Tejas7007 can address it?

File: essos/coils.py (hunk 9 above)
Comment (EstevaoMGomes) [reply essos/coils.py]: Technically yes, but I considered them to be the standard

**Comment by EstevaoMGomes** [comment]
It is fine to merge this PR first, your choice

**Comment by eduardolneto** [comment]
@EstevaoMGomes  I have merged the changes of #35 into this #30 . Let me know if you need changes or we can merge.

File: essos/objective_functions.py [hunk 10]
```diff
@@ -1,13 +1,15 @@
 import jax
+
+from ESSOS_Cópia.essos.objective_functions import field_from_dofs
```
Comment (EstevaoMGomes) [inline essos/objective_functions.py]: What is this? This causes an error in some of the examples

**Review by EstevaoMGomes — CHANGES_REQUESTED** [review]
Please test the examples. 
Many that I have tested are not working in this branch, such as `examples/comparisons_simsopt/losses.py` and `examples/paper/gradients.py`.

File: essos/objective_functions.py (hunk 10 above)
Comment (eduardolneto) [reply essos/objective_functions.py]: Removed.

Comment (eduardolneto) [reply essos/objective_functions.py]: @EstevaoMGomes  I did not changed or test the the paper and simsopt_comaprison cases, that was not the purpose of this pull request... The losses.py is fixed because there was some iteraction with the added coil cost fucntions for surface_coil separation and linking number.

However, gradients.py is failing because the entire example is making use of loss functions that are not compatiible with custom_losses (dependent of dofs and not the coils object). As I substituted those losses for the ones compatible with custom_losses in objective_functions.py, those examples need to be revamped. I think that should be done in another pull request.

**Comment by rogeriojorge** [comment]
Shall we close this pull request? Are all changes addressed in the newer PRs?

**Comment by EstevaoMGomes** [comment]
Already resolved in smaller PRs
